core/src/main/java/org/apache/iceberg/MergeAppend.java

```diff
@@ -49,6 +50,9 @@ public MergeAppend appendFile(DataFile file) {
 
   @Override
   public AppendFiles appendManifest(ManifestFile manifest) {
+    Preconditions.checkArgument(!manifest.hasExistingFiles(), "Cannot append manifest with existing files");
+    Preconditions.checkArgument(!manifest.hasDeletedFiles(), "Cannot append manifest with deleted files");
+    Preconditions.checkArgument(manifest.snapshotId() == null, "Snapshot id must be assigned during commit");
```

Comment:
Could this rewrite instead of failing? That might be an easier way to migrate to inherited snapshot IDs.

Reply:
We can add it but it will, for example, complicate the logic for clean-up as we will have to keep track of rewritten manifests. I believe users that perform manifest-level operations rely on `ManifestWriter.write(spec, file)` that is updated in this change to write nulls instead of -1. At the same time, this change forces us to migrate from `Manifest` helper class to `ManifestFile` while importing Spark tables to Iceberg.

Reply:
Agreed, we can do it in a follow-up.